05_concurrency/homework/cache/cache.go

```diff
@@ -0,0 +1,117 @@
+// Implement in-memory cache. It must be safe for concurrent usage.
+// TTL means duration while key is valid. Invalidation should happen automatically.
+// After reading a key its TTL should be increased up to current time + TTL.
+
+package cache
+
+import (
+	"sync"
+	"time"
+)
+
+type Cache interface {
+	Set(key string, value interface{}, ttl time.Duration)
+	Get(key string) (interface{}, bool)
+	Delete(key string)
+}
+
+type valueContainter struct {
+	value          interface{}
+	expirationTime time.Time
+	ttl            time.Duration
+}
+
+func (container *valueContainter) isExpired() bool {
+	return time.Now().After(container.expirationTime)
+}
+
+func (container *valueContainter) updateExpirationTime(ttl time.Duration) {
+	container.expirationTime = time.Now().Add(ttl * time.Nanosecond)
+}
+
+type SimpleCache struct {
+	storage map[string]valueContainter
+	mu      sync.Locker
+	quit    chan interface{}
+}
+
+// constructor for SimpleCache
+func NewSimpleCache(invalidationPeriod time.Duration) *SimpleCache {
+	cache := &SimpleCache{
+		storage: make(map[string]valueContainter),
+		quit:    make(chan interface{}),
+		mu:      &sync.Mutex{},
+	}
+	go runInvalidation(cache, invalidationPeriod)
+	return cache
+}
+
+func (cache *SimpleCache) Get(key string) (interface{}, bool) {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	container, ok := cache.storage[key]
+	if !ok {
+		return nil, false
+	}
+
+	if container.isExpired() {
+		return nil, false
+	}
+
+	container.updateExpirationTime(container.ttl)
+	cache.storage[key] = container
+	return container.value, true
+}
+
+func (cache *SimpleCache) Set(key string, value interface{}, ttl time.Duration) {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	tmpContainer, ok := cache.storage[key]
+
+	if !ok {
+		tmpContainer = valueContainter{}
+	}
```

Comment:
cache.storage is a map of keys by structures. Thus reading will either return the value by key or empty value of its type which will be `valueContainer`. There is no need to check whether `ok` is true or false.